### ISSUE TITLE
GCE second cluster is given first cluster's API endpoint IP

### ISSUE THREAD

**Issue by codingkriggs** [kubernetes/kops#15679]
/kind bug

Note, I have used substitutes `<bucket>` and `<project>` to hide private information in this report.

**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

```
$ kops version
Client version: 1.27.0 (git-v1.27.0)
```

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**

```
$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.3", GitCommit:"25b4e43193bcda6c7328a6d147b1fb73a33f1598", GitTreeState:"clean", BuildDate:"2023-06-14T09:53:42Z", GoVersion:"go1.20.5", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.3", GitCommit:"25b4e43193bcda6c7328a6d147b1fb73a33f1598", GitTreeState:"clean", BuildDate:"2023-06-14T09:47:40Z", GoVersion:"go1.20.5", Compiler:"gc", Platform:"linux/amd64"}
```

**3. What cloud provider are you using?**

GCE

**4. What commands did you run?  What is the simplest way to reproduce this issue?**

`$ kops create cluster cluster1.k8s.local --zones us-west1-c --state gs://<bucket>/ --project=<project> --cloud gce --yes ; echo ---------- ; kops create cluster cluster2.k8s.local --zones us-west1-c --state gs://<bucket>/ --project=<project> --cloud gce --yes`

**5. What happened after the commands executed?**

A warning line in the output reads (I modified the IPs slightly):

`W0720 05:18:57.220910    5431 create_kubecfg.go:72] Found multiple API endpoints ([34.82.159.115 35.227.35.198]), choosing arbitrarily`

when I view the contents of ~/.kube/config, I have (some bits truncated with "[...]"):

```
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: [...]
    server: https://34.82.159.115
    tls-server-name: api.internal.cluster1.k8s.local
  name: cluster1.k8s.local
- cluster:
    certificate-authority-data: [...]
    server: https://34.82.159.115
    tls-server-name: api.internal.cluster2.k8s.local
  name: cluster2.k8s.local
contexts:
- context:
    cluster: cluster1.k8s.local
    user: cluster1.k8s.local
  name: cluster1.k8s.local
- context:
    cluster: cluster2.k8s.local
    user: cluster2.k8s.local
  name: cluster2.k8s.local
current-context: cluster2.k8s.local
kind: Config
preferences: {}
users:
- name: cluster1.k8s.local
  user:
    client-certificate-data: [...]
    client-key-data: [...]
- name: cluster2.k8s.local
  user:
    client-certificate-data: [...]
    client-key-data: [...]
```

**6. What did you expect to happen?**

That warning should not have appeared, I feel, because it should know since at that point we're creating cluster2 that it shouldn't be considering the IP from cluster1. But it gets confused into thinking, when creating cluster2, that maybe the cluster1 API IP is relevant.

Also, it appears to select the first IP (cluster1's) and use it for both clusters, as shown in the config content above. Specifically, the incorrect thing here is that cluster2 is assigned cluster1's IP in the config.

Important note: this issue also manifests when executing `kops export kubecfg <cluster_name> --admin`. That is, even if I manually modified ~/.kube/config, then executed the `kops export kubecfg [...]` command, the updated config file would once again use one same IP in all locations throughout the file.

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

I will include cluster2's yaml. The yaml for cluster1 is very similar, only different in places where "cluster2" becomes "cluster1" and timestamps.

```yaml
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: "2023-07-20T05:18:02Z"
  name: cluster2.k8s.local
spec:
  api:
    loadBalancer:
      type: Public
  authorization:
    rbac: {}
  channel: stable
  cloudConfig: {}
  cloudProvider: gce
  configBase: gs://<bucket>/cluster2.k8s.local
  etcdClusters:
  - cpuRequest: 200m
    etcdMembers:
    - instanceGroup: control-plane-us-west1-c
      name: c
    manager:
      backupRetentionDays: 90
    memoryRequest: 100Mi
    name: main
  - cpuRequest: 100m
    etcdMembers:
    - instanceGroup: control-plane-us-west1-c
      name: c
    manager:
      backupRetentionDays: 90
    memoryRequest: 100Mi
    name: events
  iam:
    allowContainerRegistry: true
    legacy: false
  kubeProxy:
    enabled: false
  kubelet:
    anonymousAuth: false
  kubernetesApiAccess:
  - 0.0.0.0/0
  - ::/0
  kubernetesVersion: 1.27.3
  networking:
    cilium:
      enableNodePort: true
  nonMasqueradeCIDR: 100.64.0.0/10
  project: gtm-plutot-1
  sshAccess:
  - 0.0.0.0/0
  - ::/0
  subnets:
  - cidr: 10.0.16.0/20
    name: us-west1
    region: us-west1
    type: Public
  topology:
    dns:
      type: Private
    masters: public
    nodes: public

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2023-07-20T05:18:02Z"
  labels:
    kops.k8s.io/cluster: cluster2.k8s.local
  name: control-plane-us-west1-c
spec:
  image: ubuntu-os-cloud/ubuntu-2204-jammy-v20230606
  machineType: e2-medium
  maxSize: 1
  minSize: 1
  role: Master
  subnets:
  - us-west1
  zones:
  - us-west1-c

---

apiVersion: kops.k8s.io/v1alpha2
kind: InstanceGroup
metadata:
  creationTimestamp: "2023-07-20T05:18:03Z"
  labels:
    kops.k8s.io/cluster: cluster2.k8s.local
  name: nodes-us-west1-c
spec:
  image: ubuntu-os-cloud/ubuntu-2204-jammy-v20230606
  machineType: e2-medium
  maxSize: 1
  minSize: 1
  nodeLabels:
    cloud.google.com/metadata-proxy-ready: "true"
  role: Node
  subnets:
  - us-west1
  zones:
  - us-west1-c
```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**

Modifying the command in question (4) above with `-v 10` in the two relevant areas, the output (with some redactions for privacy) can be found here (IPs will be different since this is a different execution from above): https://gist.github.com/codingkriggs/8a02050bd333636cb13efe79f09b0d45

Jump to the line starting with "W0720 05:49:04.517824", which is near the end. Note that the file is truncated by github's UI, you'll need to view the raw file.

**9. Anything else do we need to know?**

Private topology exhibits this same behaviour.

**Comment by codingkriggs** [comment]
I can confirm that this seems to be a regression. When repeating the commands using kops v1.26.4, the content of the ~/.kube/config yaml looks correct to me:

```yaml
apiVersion: v1
clusters:
- cluster:
    [...]
    server: https://35.227.35.198
    tls-server-name: api.internal.cluster1.k8s.local
  name: cluster1.k8s.local
- cluster:
    [...]
    server: https://34.83.117.7
    tls-server-name: api.internal.cluster2.k8s.local
  name: cluster2.k8s.local
[...]
```

**Comment by justinsb** [comment]
So this should be fixed by #15709, which will be in 1.28, and by #15831 which is a (proposed) cherry-pick to 1.27

forwardingRules now support labels, so we will add a label to the forwarding rules to mark the cluster.  This is much more robust than the name-based approach we had to use before.

This does unfortunately mean that `kops export kubecfg` won't work with a cluster created with an earlier version of kOps, until you do `kops update cluster` with the new kops version.  Hopefully that is tolerable.

**Comment by codingkriggs** [comment]
/close
I am no longer around to champion this issue, so I'll fast-track it to closed.
